src/pydrex/core.py

```diff
@@ -381,3 +372,88 @@ def _get_strain_energy_enstatite(
     return dislocation_density * np.exp(
         -nucleation_efficiency * dislocation_density**2
     )
+
+
+# NOTE: Requires explicit Numba signature and must be defined last,
+# otherwise Numba 0.57.1 is not happy when CPO for multiple phases is calculated.
+# @nb.njit(
+#     "Tuple((f8[:,::1], f8))(i8, i8, f8[:,::1], f8[:,::1], f8[:,::1], f8, f8, i8)", fastmath=True
+# )
+@nb.njit
+def _get_rotation_and_strain(
+    phase,
+    fabric,
+    orientation,
+    strain_rate,
+    velocity_gradient,
+    stress_exponent,
+    deformation_exponent,
+    nucleation_efficiency,
+):
+    """Get the crystal axes rotation rate and strain energy of individual grain.
+
+    Args:
+    - `phase` (int) — ordinal number of the mineral phase
+                      see `pydrex.minerals.MineralPhase`
+    - `fabric` (int) — ordinal number of the fabric type, see `pydrex.fabric`
+    - `orientation` (array) — 3x3 orientation matrix (direction cosines)
+    - `strain_rate` (array) — 3x3 dimensionless strain rate matrix
+    - `velocity_gradient` (array) — 3x3 dimensionless velocity gradient matrix
+    - `stress_exponent` (float) — value of `p` for `dislocation_density ∝ shear_stress^p`
+    - `deformation_exponent` (float) — value of `n` for `shear_stress ∝ |deformation_rate|^(1/n)`
+    - `nucleation_efficiency (float) — parameter controlling grain nucleation
+
+    Note that "new" grains are assumed to rotate with their parent.
+
+    .. warning:: Raises zero-division errors if the vorticity is zero.
+
+    Returns a tuple with the rotation rate of the crystalline axes
+    with respect to the principal strain axes and strain energy of the grain.
+
+    """
+    crss = get_crss(phase, fabric)
+    if phase == MineralPhase.olivine:
+        slip_invariants = _get_slip_invariants_olivine(strain_rate, orientation)
+        slip_indices = np.argsort(np.abs(slip_invariants / crss))
+        slip_rates = _get_slip_rates_olivine(
+            slip_invariants,
+            slip_indices,
+            crss,
+            deformation_exponent,
+        )
+    elif phase == MineralPhase.enstatite:
+        slip_indices = np.argsort(1 / crss)
+        slip_rates = np.repeat(np.nan, 4)
+    else:
+        assert False  # Should never happen.
```

Comment:
Just a little preference here, but perhaps `raise SystemExit` instead of `assert False`? It seems more explicit to me.

Reply:
OK good idea, I will use `SystemExit(1)` with code 1 for failure for these "programmer error" errors (I'm not throwing a normal Exception here because it's not an user error).